[FIX] account: Fix cash rounding multicompany

geiercsaba opened the pull request:
Description of the issue/feature this PR addresses:
A bug occurred when the system attempted to automatically add a rounding line to an invoice in a multi-company environment. The company-dependent field was being updated without properly considering the invoice’s company context.

Current behavior before PR:
An error was raised when trying to create the rounding line because the system looked up the accounts for Company 1 instead of the invoice’s assigned company.

Desired behavior after PR is merged:
No error was raised


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/196479.png)](https://mergebot.odoo.com/odoo/odoo/pull/196479)

tsabi commented:
One of OdooTech's customer faced this issue. They have several companies in Odoo, and we setted up a global cash rounding object on the account.move model. During the running of the "invoice generation from SUB" cron job a lot of invoice runned to error, because the rounding account.move.line record contained the account_id of the company 1, however the invoice is belonged to company 2. This happened because the cron job user is belonged to company 1, and the prodit/loss account_id field on the cash rounding model is a company dependent field.

The fix can be implemented by to adjust the calling of the `_recompute_cash_rounding_lines` method, but we choosed to adjust the operating of the method instead, because this way we always guarantee to have the good company context during the accessing the fields of the cash rounding object.

We didn't tested this in later Odoos, our customer use 16.0, so maybe other version are affected too.

geiercsaba commented:
@alialfie Hey, I created a new pull requestst from the previous one. Could you please check it?

alialfie reviewed (commented):
Hi, thanks for your contribution and sorry for the delay. After reading the steps to reproduce, I'm not sure how to follow them. Could you please record a small video reproducing the issue on runbot? Thanks!

geiercsaba commented:
When I try to create an invoice under HU Company and select cash rounding (which belongs to both companies), but the active company is My Company, the system tries to use the journals of My Company for the invoice of HU Company. You can check the added automated tests for more information. Thanks!

Uploading Screen Recording 2025-02-21 at 18.34.46.mov…

alialfie reviewed (changes requested):
Thanks guys! Mostly it looks good to me, but can I ask you to please make a few changes?

1- Add a docstring to the test (similar to the one in the commit tsabi mentioned).
2- Rewrite the commit message. IMO it could be a bit clearer :) The commit tsabi mentioned again is a good example.

alialfie commented on addons/account/tests/test_invoice_tax_totals.py:
```diff
@@ -845,6 +845,44 @@ def test_recompute_cash_rounding_lines(self):
 
+    def test_recompute_cash_rounding_lines_multi_company(self):
+        """
+        Ensure that the cash rounding is determined by `move.company_id` rather than `user.company_id`.
+        OdooBot has `res.company(1)` set as the default company. The test checks that the cash correctly reflects
+        the company associated with the move, independent of the user's default company.
```
```suggestion
        Ensure that when a move is created with cash rounding that will add an invoice line, the cash rounding accounts
        will be that of the move's company and not the user's company.
```

alialfie reviewed (changes requested):
Thanks for the quick response! Almost there :smile:

geiercsaba commented:
I removed the initialization of company_data_2 since it was already created in AccountTestInvoicingCommon setup method and updated the docstring as suggested.

alialfie reviewed (approved):
Thanks! A final review will soon follow :)

FlorianGilbert commented on addons/account/tests/test_invoice_tax_totals.py:
```diff
@@ -845,6 +845,39 @@ def test_recompute_cash_rounding_lines(self):
+        Ensure that when a move is created with cash rounding that will add an invoice line, the cash rounding accounts
+        will be that of the move's company and not the user's company.
+        """
+        cash_rounding_add_invoice_line = self.env['account.cash.rounding'].with_company(
+            self.company_data["company"]).create({
```
Small indentation issue here
```suggestion
        cash_rounding_add_invoice_line = self.env['account.cash.rounding'].with_company(self.company_data["company"]).create({
```

FlorianGilbert commented on addons/account/tests/test_invoice_tax_totals.py:
```diff
@@ -845,6 +845,39 @@ def test_recompute_cash_rounding_lines(self):
         for move in moves_rounding:
             self.assertEqual(sum(move.line_ids.filtered(lambda line: line.display_type == 'rounding').mapped('balance')), moves_rounding[move])
 
+    def test_recompute_cash_rounding_lines_multi_company(self):
```
Even if we don't get the traceback, I'll personally add an assert to make sure that we got the correct account for the cash rounding line. nothing protect me and ensure me that we didn't got a silent error here

FlorianGilbert reviewed (changes requested):
LGTM, 2 remarks for the test, thank you for your contribution!

geiercsaba commented:
Done! Thanks for your feedback. Let me know if you need any further adjustments.

FlorianGilbert reviewed (commented):
Hello, may I ask you to rebase your branch as requested by runbot?

FlorianGilbert reviewed (approved):
@robodoo r+

robodoo commented:
@geiercsaba @FlorianGilbert staging failed: ci/runbot on df2b5175a231aaca5b87297783ca7a28ffaa8ae5 (view more at https://runbot.odoo.com/runbot/build/77097003)

FlorianGilbert commented:
@robodoo retry
